InformationScripting/src/graph/Graph.cpp

```diff
@@ -0,0 +1,191 @@
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#include "Graph.h"
+
+#include "InformationNode.h"
+#include "InformationEdge.h"
+
+namespace InformationScripting {
+
+QList<Graph::NodeHash> Graph::nodeHashFunctions_;
+
+Graph::~Graph()
+{
+	for (auto& e : edges_) SAFE_DELETE(e);
+	for (auto& n : nodes_) SAFE_DELETE(n);
+}
+
+InformationNode* Graph::add(InformationNode* node)
+{
+	Q_ASSERT(!nodeHashFunctions_.empty());
+
+	InformationNode* existingNode = findNode(node);
+
+	if (existingNode && existingNode != node)
+	{
+		mergeNodes(existingNode, node);
+		SAFE_DELETE(node);
+	}
+	else
+	{
+		nodes_[hashValueOf(node)] = node;
+		existingNode = node;
+	}
+	return existingNode;
+}
+
+InformationEdge* Graph::addDirectedEdge(InformationNode* from, InformationNode* to, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(from == findNode(from));
+	Q_ASSERT(to == findNode(to));
+	for (auto edge : edges_)
+	{
+		if (edge->from() == from && edge->to() == to && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(from, to, name);
+	edges_.push_back(edge);
+	return edge;
+}
+
+InformationEdge* Graph::addEdge(InformationNode* a, InformationNode* b, const QString& name)
+{
+	// We only allow existing nodes:
+	Q_ASSERT(a == findNode(a));
+	Q_ASSERT(b == findNode(b));
+	for (auto edge : edges_)
+	{
+		if ((edge->from() == a || edge->from() == b) && (edge->to() == b || edge->to() == a) && edge->name() == name)
+		{
+			auto existingEdge = edge;
+			Q_ASSERT(!existingEdge->isDirected());
+			existingEdge->incrementCount();
+			return existingEdge;
+		}
+	}
+	auto edge = new InformationEdge(a, b, name, InformationEdge::Orientation::Undirected);
+	edges_.push_back(edge);
+	return edge;
+}
+
+void Graph::remove(InformationNode* node)
+{
+	Q_ASSERT(node);
+	auto hash = hashValueOf(node);
+	auto nodeIt = nodes_.find(hash);
+	// If the node isn't in the graph we don't have to do anything:
+	if (nodeIt == nodes_.end()) return;
+
+	// It could be that node points to a different node with the same hash.
+	// TODO in this case we might also want to merge the infos?
+	auto ownedNode = node;
+	if (ownedNode != *nodeIt) ownedNode = *nodeIt;
+
+	for (auto edgeIt = edges_.begin(); edgeIt != edges_.end();)
+	{
+		auto edge = *edgeIt;
+		if (edge->from() == ownedNode || edge->to() == ownedNode)
+		{
+			SAFE_DELETE(edge);
+			edgeIt = edges_.erase(edgeIt);
+		}
+		else
+		{
```

Comment:
Hm true it's from the Qt style: http://wiki.qt.io/Qt_Coding_Style (search for Exception 2). I can remove it if you prefer